tools/metadatastore-upgrade/main.go

```diff
@@ -0,0 +1,152 @@
+			result.Spec.Template.Spec.Containers[0].Args = containerArgs
+			_, updateErr := deploymentsClient.Update(result)
+			if updateErr != nil {
+				return updateErr
+			}
+
+			result, getErr = deploymentsClient.Get( ML_METADATA_DEPLOYMENT,  metav1.GetOptions{})
+			if getErr != nil {
+				return fmt.Errorf("Failed to get latest version of Deployment after update: %v", getErr)
+			}
+
+			waitTime := 0
+			updateSuccessful := false
+			for waitTime < MAX_WAIT_TIME && !updateSuccessful {
+				if *result.Spec.Replicas != result.Status.ReadyReplicas {
+					time.Sleep(time.Second)
+					waitTime++
+				} else {
+					updateSuccessful = true
+				}
+		}
+
+		if !updateSuccessful {
+			return fmt.Errorf("Updated deployment failed to reach running state")
+		}
+
+		return nil
+	})
+}
+
+
+func main() {
+	flag.Parse()
+	log.SetOutput(os.Stdout)
+
+	if *imageTag == "" {
+		log.Printf("Missing image_tag flag")
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	if *kubeconfigPath == "" {
+		if home := homedir.HomeDir(); home != "" {
+			*kubeconfigPath = filepath.Join(home, ".kube", "config")
+		}
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfigPath)
+	if err != nil {
+		log.Printf("Error reading kubeconfig file: %v", err)
+		os.Exit(1)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Printf("Error setting up client auth: %v", err)
+		os.Exit(1)
+	}
+
+	deploymentsClient := clientset.AppsV1().Deployments(*deploymentNamespace)
+
+	oldDeployment, getErr := deploymentsClient.Get( ML_METADATA_DEPLOYMENT,  metav1.GetOptions{})
+	if getErr != nil {
+		log.Printf("Failed to get old Deployment: %v", getErr)
+		os.Exit(1)
+	}
+
+	originalImage := oldDeployment.Spec.Template.Spec.Containers[0].Image
+	originalContainerArgs := oldDeployment.Spec.Template.Spec.Containers[0].Args
+
+	newImage := ML_METADATA_IMAGE + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s to Image: %s", *deploymentNamespace, newImage)
+	updateErr := UpdateDeployment(deploymentsClient, newImage, append(originalContainerArgs, UPGRADE_FLAG))
+	if updateErr == nil {
+	  log.Printf("MetadataStore successfully upgraded to Image: %s", newImage)
+	  log.Printf("Cleaning up Upgrade")
+		upgradeCleanupErr := 	UpdateDeployment(deploymentsClient, newImage, originalContainerArgs)
+		// In a highly unlikely scenario upgrade cleanup can fail.
+		if upgradeCleanupErr != nil {
+			log.Printf("Upgrade cleanup failed. Likely MetadataStore is in a functioning state. Please contact Kubeflow community for verification")
+			os.Exit(1)
+		}
+	} else {
+		log.Printf("MetadataStore Upgrade failed. Reverting back to Old Image")
+		revertUpdateErr := UpdateDeployment(deploymentsClient, originalImage, originalContainerArgs)
+		if revertUpdateErr != nil {
+			log.Printf("Revert failed. Please contact Kubeflow community")
+			os.Exit(1)
+		} else {
+			log.Printf("Revert succeeded. MetadataStore unchanged.")
+			os.Exit(1)
```

Comment:
Is this os.Exit needed here?

Reply:
Yes because the reason for running the tool was un-successful. It would be better to return failure in cases where this tool is used as part of a script.